# ¡Hola Juan!

Mi nombre es Ezequiel Ferrario, soy code reviewer en Tripleten y tengo el agrado de revisar el proyecto que entregaste.

Para simular la dinámica de un ambiente de trabajo, si veo algún error, en primer instancia solo los señalaré, dándote la oportunidad de encontrarlos y corregirlos por tu cuenta. En un trabajo real, el líder de tu equipo hará una dinámica similar. En caso de que no puedas resolver la tarea, te daré una información más precisa en la próxima revisión.

Encontrarás mis comentarios más abajo - **por favor, no los muevas, no los modifiques ni los borres**.

¿Cómo lo voy a hacer? Voy a leer detenidamente cada una de las implementaciones que has llevado a cabo para cumplir con lo solicitado. Verás los comentarios de esta forma:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Si todo está perfecto.
</div>


<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta. Se aceptan uno o dos comentarios de este tipo en el borrador, pero si hay más, deberá hacer las correcciones. Es como una tarea de prueba al solicitar un trabajo: muchos pequeños errores pueden hacer que un candidato sea rechazado.
</div>

<div class="alert alert-block alert-danger">

<b>Comentario del revisor</b> <a class="tocSkip"></a>

Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>

Puedes responderme de esta forma:

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>

Hola, muchas gracias por tus comentarios y la revisión.
</div>

¡Empecemos!

# Déjame escuchar la música

# Contenido <a id='back'></a>

* [Introducción](#Introducción)
    * [Objetivo](#Objetivo)
    * [Etapas](#Etapas)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

<div class="alert alert-block alert-warning">

<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy bien la tabla de contenidos pero debe estar linkeada a las secciones (al clickear debe llevarnos a esa seccion) de esta menera es mas facil desplazarse.

Como consejo, si realizas bien todas las secciones (con su respectivo #) podes generarlo automáticamente desde jupyter lab. Para hacerlo, en la pestaña de herramientas de jupyter lab clickeas en el **botón de los puntos y barras**  (Table of contents) te generara automáticamente una tabla de contenidos linkeable y estética. A la **derecha** del botón "Validate"
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

Corregido.</div>

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo
Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.


### Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.








[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Necesitarás `pandas`, así que impórtalo.

In [1]:
# Importar pandas
import pandas as pd
# Le coloco el alias "pd" como estándar de la industria, aunque el ejercicio no lo pida directamente

Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [2]:
# Leer el archivo y almacenarlo en df
df = pd.read_csv("/datasets/music_project_en.csv")
# Nos apoyamos del método read_csv de pandas para abrir el dataset en formato CSV.

Muestra las 10 primeras filas de la tabla:

In [3]:
# Obtener las 10 primeras filas de la tabla df
df.head(10)
# Usamos el método head() que muestra por defecto las 5 primeras filas del Dataset
# Pasamos como argumento el número 10 que es la cantidad de filas que nos propone el ejercicio

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

Obtén la información general sobre la tabla con un comando. Conoces el método que muestra la información general que necesitamos.

In [4]:
# Obtener la información general sobre nuestros datos
df.info()
# El método info() retorna el número de filas, el número de columnas.
# Además nos da un resumen de las columnas con su índice, nombre, valores no nulos y tipo de datos que contiene
# Finalmente el uso de memoria del DataFrame

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. `Detecta el tercer problema por tu cuenta y descríbelo aquí`.

Se me ocurre que el tercer problema puede ser que faltan algunos espacios en los encabezados, además que no son lo suficientemente explicativos respecto al contenido que encontramos. Por ejemplo, la primera columna podría ser `user_id` y la última columna podría ser llamada `day_of_the_week`

### Escribe observaciones de tu parte. Estas son algunas de las preguntas que pueden ser útiles: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?`
Todos son de tipo 'object', es decir, en este caso en específico de este DataFrame es que son strings. Tiene sentido principalmente debido al tipo de manipulación que se pueden lograr con los datos. 

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`
El curso me ha enseñado hasta este punto que nunca van a estar de más poder tener acceso a más información. Sin embargo, parece ser que lo que se nos brinda en este momento es justamente lo que necesitamos para llegar a conclusiones importantes. 

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`
Las primeras exploraciones en los datos me llevan a dar una respuesta afirmativa. A esto llegué abriendo una celda antes de contestar estas preguntas y preguntando por los siguientes métodos a Pandas. En el caso de valores ausentes se encuentran 1343 sólo en la columna de Track, 7567 en la columna de artist y 1198 en la columna de genre. Tenemos 3826 filas duplicadas.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Buenas observaciones.</div>

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>

Hola, muchas gracias por tus comentarios y la revisión.
</div>

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [5]:
# Muestra los nombres de las columnas
df.columns

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

Anteriormente, aprendiste acerca de la forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [6]:
# Bucle en los encabezados poniendo todo en minúsculas
new_col_names = []
for old_name in df.columns:
    lower_name = old_name.lower()
    new_col_names.append(lower_name)
df.columns = new_col_names
print(df.columns)
# Usamos el método lower() para asegurarnos que toda la cadena se convierta a minúsculas.

Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')


Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [7]:
# Bucle en los encabezados eliminando los espacios
new_col_names = []
for old_name in df.columns:
    stripped_name = old_name.strip()
    new_col_names.append(stripped_name)
df.columns = new_col_names
print(df.columns)
# Usamos el método strip() para eliminar espacios innecesarios al comienzo y al final del nombre

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [8]:
# Cambiar el nombre de la columna "userid"
df.rename(columns ={"userid": "user_id", "day": "day_of_the_week"}, inplace = True)
print(df.columns)
# Cambiamos directamente el nombre de dos columnas con el método rename()

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time',
       'day_of_the_week'],
      dtype='object')


Comprueba el resultado. Muestra los encabezados una vez más:

In [9]:
# Comprobar el resultado: la lista de encabezados
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time',
       'day_of_the_week'],
      dtype='object')

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy bien.</div>

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>

Hola, muchas gracias por tus comentarios y la revisión.
</div>

[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [10]:
# Calcular el número de valores ausentes
df.isna().sum()
# El método isna() nos devuelve True o False dependiendo si el valor de determinada celda está ausente o no respectivamente
# Sumamos todos los valores True (verdaderos) que encontró isna() con el método sum()

user_id               0
track              1343
artist             7567
genre              1198
city                  0
time                  0
day_of_the_week       0
dtype: int64

No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [11]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
col_to_replace = ["track", "artist", "genre"]
for col in col_to_replace:
    df[col].fillna("unknown", inplace = True)
# Iteramos sobre la lista de columnas a reemplazar y usamos el método fillna() para reemplazar los valores ausentes.

Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

In [12]:
# Contar valores ausentes
df.isna().sum()

user_id            0
track              0
artist             0
genre              0
city               0
time               0
day_of_the_week    0
dtype: int64

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Bien utilizada la funcion para remplazar los nulos.</div>

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>

Muchas gracias.
</div>

[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [13]:
# Contar duplicados explícitos
df.duplicated().sum()
# El método duplicated() nos devuelve una lista con valores True o False si encuentra duplicados
# Luego con el método sum() sumamos todos los True

3826

Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [14]:
# Eliminar duplicados explícitos
df.drop_duplicates().reset_index()
# Elimino los duplicados con el método .drop_duplicates()
# Luego, con .reset_index() me aseguro de ver la diferencia en la aplicación del método
df = df.drop_duplicates().reset_index(drop = True)
# Una vez encontrada la diferencia, actualizo los índices con el argumento drop = True

Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [15]:
# Comprobar de nuevo si hay duplicados
df.duplicated().sum()

0

Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [1]:
# Inspeccionar los nombres de géneros únicos
list_of_genres = df["genre"].unique()
print(sorted(list_of_genres))
# Guardo los valores únicos de la columna en la variable list_of_genres usando el método unique()
# Luego los ordeno en orden alfabético para encontrar los duplicados implícitos más fácilmente con la función sorted()

NameError: name 'df' is not defined

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy bien.

Como consejo, contando la cantidad de elementos unicos que tenemos (antes y despues) es  un parametro para verificar si el error esta corregido o no.

Si tengo menos registros que antes estara corregido.</div>

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>

Entiendo, claro, tienes toda la razón.
</div>

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [17]:
# Función para reemplazar duplicados implícitos
def replace_wrong_genres(wrong_genres, correct_genre):
    for wrong in wrong_genres:
        df["genre"].replace(wrong, correct_genre, inplace = True)
# Itero la función sobre la columna necesaria y aplico el método replace() para modificar cada valor incorrecto inplace.

Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [18]:
# Eliminar duplicados implícitos
bad_spell = ["hip", "hop", "hip-hop"]
replace_wrong_genres(bad_spell, "hiphop")
# Creo una lista con todos los posibles nombres mal escritos
# Luego aplico la función para cambiar cada nombre mal escrito por el correcto

Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [19]:
# Comprobación de duplicados implícitos
list_of_genres = df["genre"].unique()
print(sorted(list_of_genres))

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

`Describe brevemente lo que has notado al analizar duplicados, cómo abordaste sus eliminaciones y qué resultados obtuviste.`
A pesar de las grandes capacidades de Pandas, siento que los duplicados implícitos es la parte que más tiempo toma, posiblemente porque es de los pasos donde se vuelve necesaria una interpretación humana, pues en la mayoría de los casos, son errores que surgieron por humanos. También creo que es el primer momento de todo el análisis en el que es necesario tener contexto del DataFrame, poder interpretar los géneros de forma correcta, es decir, en realidad podría existir un género llamado `hip` además del `hiphop`, por ejemplo.

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Esta muy bueno lo que comentas.
    
Recorda que tratamos de simular un espacio profesional donde los proyectos son dirigidos al cliente. Por lo que estos aportados de observacion o conclusion intermedia deberan usarse para explicar que cambios se realizaron en el dataset y porque, ademas de dejar en claro si se encontraron errores.
    
Esto se hace para recapitular lo realizado y darle un panorama al cliente de que problemas se podrian mejorar e informar lo realizado
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

Corregido. Te quedo excelente las observaciones.</div>

Hubo varias cosas que se revisaron antes de comenzar un análisis directo de la hipótesis. Se comenzó actualizando el nombre de las columnas: resulta que nos encontramos con unos nombres que no eran lo suficientemente explicativos de su contenido, además de algunas fallas de sintaxis como que encontramos unos espacios antes y después de los nombres, encontramos encabezados con usos de mayúsculas aleatorias.

Luego, sería importante por el rigor de la información, poder entender la razón de la falta de valores en las columnas de canciones, artistas y géneros. Esta información podría resultar útil si intentamos generar hipótesis nuevas a partir de esta primera conclusión. Las reemplazamos con la palabra `unknown` para que no obstaculice nuestro análisis, pero no lo podíamos dejar pasar por alto.

Finalmente, se revisaron algunos valores concretos de la columna `genre` en la que encontramos que habían errores en la escritura de algunos géneros.

[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [20]:
# Contar las canciones reproducidas en cada ciudad
df.groupby("city")["track"].count()

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64

`Comenta tus observaciones aquí`
Es evidente que se han escuchado muchas más canciones en Springfield que en Shelbyville. La diferencia es más del doble.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy bien.
</div>

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [21]:
# Calcular las canciones reproducidas en cada uno de los tres días
df.groupby("day_of_the_week")["track"].count()

day_of_the_week
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64

`Comenta tus observaciones aquí`
No hay una diferencia evidente entre los días de la semana. Claramente la mayor cantidad de canciones se hizo el viernes, seguido del lunes, pero no es una gran diferencia.

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

¿Con respecto al miercoles?. ¿Se podria dar alguna explicacion a esto?.</div>

El miércoles resulta ser un día muy especial en nuestro análisis. Una vez analizamos las diferencias en las ciudades esto se amplía. Resulta que, en general, es el día que menos música se escuchó, pero en realidad esto sólo ocurre en la ciudad de Springfield, que cabe aclarar, es la ciudad en la que más datos encontramos, por lo que hace que se baje considerablemente la estadística. Pero en detalle, en nuestra segunda ciudad, Shelbyville, éste es justamente el día donde más registros tenemos de escucha de música. 

Entonces ahora podríamos avanzar en una investigación más intensiva que nos lleve a comprender por qué se escucha menos música los miércoles en Springfield, pero se escucha mucha más música el mismo día en Shelbyville.

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [22]:
def number_tracks(day, city):
    df_day_filtered = df.loc[df["day_of_the_week"] == day]
    df_city_filtered = df_day_filtered.loc[df_day_filtered["city"] == city]
    user_count = df_city_filtered["user_id"].count()
    return user_count

Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [23]:
# El número de canciones reproducidas en Springfield el lunes
print(number_tracks("Monday", "Springfield"))

15740


In [24]:
# El número de canciones reproducidas en Shelbyville el lunes
print(number_tracks("Monday", "Shelbyville"))

5614


In [25]:
# El número de canciones reproducidas en Springfield el miércoles
print(number_tracks("Wednesday", "Springfield"))

11056


In [26]:
# El número de canciones reproducidas en Shelbyville el miércoles
print(number_tracks("Wednesday", "Shelbyville"))

7003


In [27]:
# El número de canciones reproducidas en Springfield el viernes
print(number_tracks("Friday", "Springfield"))

15945


In [28]:
# El número de canciones reproducidas en Shelbyville el viernes
print(number_tracks("Friday", "Shelbyville"))

5895


In [29]:
days = [
    ["Monday", number_tracks("Monday", "Springfield"), number_tracks("Monday", "Shelbyville")],
    ["Wednesday", number_tracks("Wednesday", "Springfield"), number_tracks("Wednesday", "Shelbyville")],
    ["Friday", number_tracks("Friday", "Springfield"), number_tracks("Friday", "Shelbyville")],
]
cities = ["day", "Springfield", "Shelbyville"]
df_final = pd.DataFrame(data = days, columns = cities)
df_final

day  Springfield  Shelbyville
0     Monday        15740         5614
1  Wednesday        11056         7003
2     Friday        15945         5895

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Excelente idea realizar un dataframe para una mejor visualizacion.</div>

Me pareció útil poder condensar los seis resultados de una forma ordenada creando y mostrando un nuevo DataFrame en Pandas. De esa forma cualquier persona puede acceder directamente a los resultados

**Conclusiones**

`Comenta si la hipótesis es correcta o se debe rechazar. Explica tu razonamiento.`
Acepto la hipótesis completamente debido a que, como lo muestro en el DataFrame `df_final` no existe ninguna relación entre ninguna de las celdas. No se puede encontrar una relación ni un las columnas, es decir no podemos decir que se siga un patrón respecto a la ciudades, porque en Springfield se escucha más música los viernes, mientras en Shelbyville lo hacen los miércoles. Tampoco podemos construir ningún patrón en relación a las filas, porque a pesar que en Springfield haya más consumo que en Shelbyville, no hay una relación clara entre sus valores, por ejemplo el máximo de Springfield fue el viernes con 15945, mientras el máximo de Shelbyville fue el miércoles con 7003. Ahora, el mínimo de Springfield fue el miércoles y el de SHelbyville fue el lunes.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy bien.
</div>

[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

`Resume aquí tus conclusiones sobre la hipótesis.`

Acepto completamente la hipótesis. No se puede establecer una relación matemática directa entre los valores de las ciudades ni los valores de los días.

### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

<div class="alert alert-block alert-danger ">
<b>Comentario general</b> <a class="tocSkip"></a>

Juan, presentaste un proyecto muy bueno.

Me gusto mucho como utilizaste el codigo a traves de este trabajo y aun mas me gusto tus analisis que fueron minucios y explicativos, bajaste a tierra muchisima informacion.

El trabajo esta muy bien y queda un detalle para aprobar el mismo.

Espero tu correccion, saludos.</div>

<div class="alert alert-block alert-success ">
<b>Comentario general #2</b> <a class="tocSkip"></a>

Juan, los detalles que faltaban fueron corregidos muy bien.

Te queda un proyecto muy bien realizado y completo por lo que pasa a estar **aprobado**.

Exitos en lo que viene, saludos.</div>

[Volver a Contenidos](#back)